Possible data race in tsi1 package

## What

CircleCI reported the following data race running the `TestShard_WritePoints_FieldConflictConcurrent` test:

```
Failed
=== RUN   TestShard_WritePoints_FieldConflictConcurrent
==================
WARNING: DATA RACE
Write at 0x00c000a00090 by goroutine 336:
  runtime.mapassign_faststr()
      /usr/local/go/src/runtime/map_faststr.go:202 +0x0
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).execDeleteTagValueEntry()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:663 +0x48e
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).execEntry()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:633 +0x87
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).DeleteTagValue()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:510 +0x1d9
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Partition).DropMeasurement.func2()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:610 +0x1d5
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Partition).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:611 +0x2ab
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement.func1()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:609 +0x84

Previous read at 0x00c000a00090 by goroutine 439:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*logTagKey).TagValueIterator()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:1365 +0x8a
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.tagKeyMergeElem.TagValueIterator()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/tsi1.go:271 +0x101
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*tagKeyMergeElem).TagValueIterator()
      <autogenerated>:1 +0x76
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Partition).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:604 +0x1b7
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement.func1()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:609 +0x84

Goroutine 336 (running) created at:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:603 +0x17d
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurementIfSeriesNotExist()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:841 +0xca
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).deleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1734 +0xdf7
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRangeWithPredicate()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1495 +0x904
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1387 +0x91
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1832 +0x324
  github.com/influxdata/influxdb/v2/tsdb.(*Shard).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/shard.go:743 +0xb0
  github.com/influxdata/influxdb/v2/tsdb_test.TestShard_WritePoints_FieldConflictConcurrent.func2()
      /go/src/github.com/influxdata/influxdb/tsdb/shard_test.go:471 +0x104

Goroutine 439 (running) created at:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:603 +0x17d
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurementIfSeriesNotExist()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:841 +0xca
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).deleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1734 +0xdf7
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRangeWithPredicate()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1495 +0x904
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1387 +0x91
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1832 +0x324
  github.com/influxdata/influxdb/v2/tsdb.(*Shard).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/shard.go:743 +0xb0
  github.com/influxdata/influxdb/v2/tsdb_test.TestShard_WritePoints_FieldConflictConcurrent.func1()
      /go/src/github.com/influxdata/influxdb/tsdb/shard_test.go:455 +0x104
==================
==================
WARNING: DATA RACE
Write at 0x00c00036fdc8 by goroutine 336:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).execDeleteTagValueEntry()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:663 +0x4af
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).execEntry()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:633 +0x87
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).DeleteTagValue()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:510 +0x1d9
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Partition).DropMeasurement.func2()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:610 +0x1d5
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Partition).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:611 +0x2ab
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement.func1()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:609 +0x84

Previous read at 0x00c00036fdc8 by goroutine 439:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*logTagKey).TagValueIterator()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:1366 +0x1a7
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.tagKeyMergeElem.TagValueIterator()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/tsi1.go:271 +0x101
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*tagKeyMergeElem).TagValueIterator()
      <autogenerated>:1 +0x76
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Partition).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:604 +0x1b7
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement.func1()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:609 +0x84

Goroutine 336 (running) created at:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:603 +0x17d
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurementIfSeriesNotExist()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:841 +0xca
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).deleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1734 +0xdf7
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRangeWithPredicate()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1495 +0x904
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1387 +0x91
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1832 +0x324
  github.com/influxdata/influxdb/v2/tsdb.(*Shard).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/shard.go:743 +0xb0
  github.com/influxdata/influxdb/v2/tsdb_test.TestShard_WritePoints_FieldConflictConcurrent.func2()
      /go/src/github.com/influxdata/influxdb/tsdb/shard_test.go:471 +0x104

Goroutine 439 (running) created at:
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:603 +0x17d
  github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*Index).DropMeasurementIfSeriesNotExist()
      /go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:841 +0xca
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).deleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1734 +0xdf7
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRangeWithPredicate()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1495 +0x904
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteSeriesRange()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1387 +0x91
  github.com/influxdata/influxdb/v2/tsdb/engine/tsm1.(*Engine).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1832 +0x324
  github.com/influxdata/influxdb/v2/tsdb.(*Shard).DeleteMeasurement()
      /go/src/github.com/influxdata/influxdb/tsdb/shard.go:743 +0xb0
  github.com/influxdata/influxdb/v2/tsdb_test.TestShard_WritePoints_FieldConflictConcurrent.func1()
      /go/src/github.com/influxdata/influxdb/tsdb/shard_test.go:455 +0x104
==================
    testing.go:1042: race detected during execution of test
--- FAIL: TestShard_WritePoints_FieldConflictConcurrent (2.10s)
```

--- a/tsdb/index/tsi1/log_file.go
+++ b/tsdb/index/tsi1/log_file.go
@@ -368,7 +368,7 @@ func (f *LogFile) TagKeyIterator(name []byte) TagKeyIterator {
 	for _, k := range mm.tagSet {
 		a = append(a, k)
 	}
-	return newLogTagKeyIterator(a)
+	return newLogTagKeyIterator(f, a)
 }
 
 // TagKey returns a tag key element.
@@ -766,6 +766,7 @@ func (f *LogFile) createMeasurementIfNotExists(name []byte) *logMeasurement {
 	mm := f.mms[string(name)]
 	if mm == nil {
 		mm = &logMeasurement{
+			f:      f,
 			name:   name,
 			tagSet: make(map[string]logTagKey),
 			series: make(map[uint64]struct{}),
@@ -1204,6 +1205,7 @@ func (mms *logMeasurements) bytes() int {
 }
 
 type logMeasurement struct {
+	f         *LogFile
 	name      []byte
 	tagSet    map[string]logTagKey
 	deleted   bool
@@ -1304,7 +1306,7 @@ func (m *logMeasurement) Deleted() bool { return m.deleted }
 func (m *logMeasurement) createTagSetIfNotExists(key []byte) logTagKey {
 	ts, ok := m.tagSet[string(key)]
 	if !ok {
-		ts = logTagKey{name: key, tagValues: make(map[string]logTagValue)}
+		ts = logTagKey{f: m.f, name: key, tagValues: make(map[string]logTagValue)}
 	}
 	return ts
 }
@@ -1341,6 +1343,7 @@ func (itr *logMeasurementIterator) Next() (e MeasurementElem) {
 }
 
 type logTagKey struct {
+	f         *LogFile
 	name      []byte
 	deleted   bool
 	tagValues map[string]logTagValue
@@ -1362,10 +1365,13 @@ func (tk *logTagKey) Key() []byte   { return tk.name }
 func (tk *logTagKey) Deleted() bool { return tk.deleted }
 
 func (tk *logTagKey) TagValueIterator() TagValueIterator {
+	tk.f.mu.RLock()
 	a := make([]logTagValue, 0, len(tk.tagValues))
 	for _, v := range tk.tagValues {
 		a = append(a, v)
 	}
+	tk.f.mu.RUnlock()
+
 	return newLogTagValueIterator(a)
 }
 
@@ -1459,13 +1465,14 @@ func (a logTagValueSlice) Less(i, j int) bool { return bytes.Compare(a[i].name, 
 
 // logTagKeyIterator represents an iterator over a slice of tag keys.
 type logTagKeyIterator struct {
+	f *LogFile
 	a []logTagKey
 }
 
 // newLogTagKeyIterator returns a new instance of logTagKeyIterator.
-func newLogTagKeyIterator(a []logTagKey) *logTagKeyIterator {
+func newLogTagKeyIterator(f *LogFile, a []logTagKey) *logTagKeyIterator {
 	sort.Sort(logTagKeySlice(a))
-	return &logTagKeyIterator{a: a}
+	return &logTagKeyIterator{f: f, a: a}
 }
 
 // Next returns the next element in the iterator.

